src/transport_portals4.h

```diff
@@ -451,29 +458,42 @@ shmem_transport_portals4_put_nb_internal(void *target, const void *source, size_
         shmem_transport_portals4_long_pending = 1;
 #endif
     } else {
-        shmem_transport_portals4_long_frag_t *long_frag;  
+        shmem_transport_portals4_long_frag_t *long_frag;
+        ptl_handle_md_t md;
 
-        SHMEM_MUTEX_LOCK(shmem_internal_mutex_ptl4_event_slots);
-        while (0 >= --shmem_transport_portals4_event_slots) {
-            shmem_transport_portals4_event_slots++;
-            shmem_transport_portals4_drain_eq();
-        }
-        SHMEM_MUTEX_UNLOCK(shmem_internal_mutex_ptl4_event_slots);
+        shmem_internal_assert(len <= shmem_transport_portals4_max_msg_size);
 
-        long_frag = (shmem_transport_portals4_long_frag_t*)
-            shmem_free_list_alloc(shmem_transport_portals4_long_frags);
-        if (NULL == long_frag) { RAISE_ERROR(-1); }
+        /* If user requested completion notification, create a frag object and
+         * append the completion pointer */
+        if (NULL != completion) {
+            md = shmem_transport_portals4_put_event_md_h;
 
-        assert(long_frag->frag.type == SHMEM_TRANSPORT_PORTALS4_TYPE_LONG);
-        assert(long_frag->reference == 0);
-        long_frag->completion = completion;
+            SHMEM_MUTEX_LOCK(shmem_internal_mutex_ptl4_event_slots);
+            while (0 >= --shmem_transport_portals4_event_slots) {
+                shmem_transport_portals4_event_slots++;
+                shmem_transport_portals4_drain_eq();
+            }
+            SHMEM_MUTEX_UNLOCK(shmem_internal_mutex_ptl4_event_slots);
 
-        /* NOTE-MT: Frag mutex is not needed here because the frag doesn't get
-         * exposed to other threads until the PtlPut. */
-        (*(long_frag->completion))++;
-        long_frag->reference++;
+            long_frag = (shmem_transport_portals4_long_frag_t*)
+                shmem_free_list_alloc(shmem_transport_portals4_long_frags);
+            if (NULL == long_frag) { RAISE_ERROR(-1); }
 
-        ret = PtlPut(shmem_transport_portals4_put_event_md_h,
+            shmem_internal_assert(long_frag->frag.type == SHMEM_TRANSPORT_PORTALS4_TYPE_LONG);
+            shmem_internal_assert(long_frag->reference == 0);
+            long_frag->completion = completion;
+
+            /* NOTE-MT: Frag mutex is not needed here because the frag doesn't get
+             * exposed to other threads until the PtlPut. */
+            (*(long_frag->completion))++;
+            long_frag->reference++;
+
+        } else {
+            md = shmem_transport_portals4_put_cntr_md_h;
+            long_frag = NULL;
+        }
+
+        ret = PtlPut(md,
```

Comment:
@kseager, I reviewed this commit again and it looks ok.  In Portals 4, volatile is a property of the MD.  Before this change, we had two MDs available for issuing puts: (1) volatile and (2) full event.  This is because OpenSHMEM 1.2 only had blocking puts, so we either issued volatile/inject puts (that are locally completed immediately) or we asked for a full event to set the user's completion handle.  I had to add another MD to capture counting events for nonblocking puts for scenarios where we don't want a full event back (e.g. user didn't give a completion handle).  The ```put_cntr_md_h``` MD uses the same counter as the volatile MD, so there were no changes needed to quiet.